File: app/src/main/java/org/dslul/openboard/inputmethod/latin/utils/RunInLocale.java
/*
 * Copyright (C) 2013 The Android Open Source Project
 * modified
 * SPDX-License-Identifier: Apache-2.0 AND GPL-3.0-only
 */

package org.dslul.openboard.inputmethod.latin.utils;

import android.content.res.Configuration;
import android.content.res.Resources;

import java.util.Locale;

public abstract class RunInLocale<T> {
    private static final Object sLockForRunInLocale = new Object();

    protected abstract T job(final Resources res);

    /**
     * Execute {@link #job(Resources)} method in specified system locale exclusively.
     *
     * @param res the resources to use.
     * @param newLocale the locale to change to. Run in system locale if null.
     * @return the value returned from {@link #job(Resources)}.
     */
    public T runInLocale(final Resources res, final Locale newLocale) {
        synchronized (sLockForRunInLocale) {
            final Configuration conf = res.getConfiguration();
            if (newLocale == null || newLocale.equals(conf.locale)) {
                return job(res);
            }
            final Locale savedLocale = conf.locale;
            try {
                conf.locale = newLocale;
                res.updateConfiguration(conf, null);
                return job(res);
            } finally {
                conf.locale = savedLocale;
                res.updateConfiguration(conf, null);
            }
        }
    }
}

File: app/src/main/java/org/oscar/kb/latin/utils/RunInLocale.kt
// SPDX-License-Identifier: GPL-3.0-only
package org.oscar.kb.latin.utils

import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import java.util.Locale

fun <T> runInLocale(context: Context, locale: Locale, run: (Context) -> T): T {
    val config = Configuration(context.resources.configuration)
    config.setLocale(locale)
    val localeContext = context.createConfigurationContext(config)
    return run(localeContext)
}

// slower than context and deprecated, but still better than original style
fun <T> runInLocale(resources: Resources, locale: Locale, run: (Resources) -> T): T {
    val config = Configuration(resources.configuration)
    config.setLocale(locale)
    @Suppress("deprecation") val res = Resources(resources.assets, resources.displayMetrics, config)
    return run(res)
}
